net/conn_pool.go

```diff
@@ -1,69 +1,182 @@
+	NumberOfUsers() int64
+	// Done reports to the ConnPool that you are done using this ConnPoolObject.
+	// This should be done after ALL operations against the object are resolved
+	Done()
+}
+
+// connNode is a circulary linked list of connection
+// instead of using an array where the earlier connections would
+// get much higher load, we'll just run around the loop, so we balance
+// in true round-robin fashion
+type connNode struct {
+	obj ConnPoolObject
+
+	numberUsers int64
+	loopback    chan<- *connNode
+
+	// only for use if obj is ConnPoolObjectCanMux
+	index int
+	value int
+
+	deleted uint32
+}
+
+func (cn *connNode) incrementUsers() int64 {
+	return atomic.AddInt64(&cn.numberUsers, 1)
+}
+
+func (cn *connNode) decrementUsers() int64 {
+	return atomic.AddInt64(&cn.numberUsers, -1)
+}
+
+func (cn *connNode) ConnPoolObject() ConnPoolObject {
+	return cn.obj
+}
+
+func (cn *connNode) NumberOfUsers() int64 {
+	return atomic.LoadInt64(&cn.numberUsers)
+}
+
+func (cn *connNode) Done() {
+	cn.decrementUsers()
+	cn.loopback <- cn
+}
+
 // ConnPool is a fast concurrency safe connection pool structure
 type ConnPool interface {
 	// Insert returns a no error false case only when we try to insert
 	// beyond our max connections limit
 	Insert(ConnPoolObject) (bool, error)
-	Get() ConnPoolObject
+	Get() ConnPoolContext
+}
+
+// connPool is designed to be a speedy connection pool
+// Extensive testing for race conditions is a priority.
+// DO NOT edit without updating or checking against existing tests
+type connPool struct {
+	numConns        int64
+	numMuxableConns int64
+	maxConns        int64
+
+	logger           *logrus.Entry
+	delConnCh        chan *connNode
+	noActivityConnCh chan *connNode
+	loopbackConnCh   chan *connNode
+
+	muxHeap            muxableConnHeap
+	muxHeapEmptyMutex  sync.RWMutex // protects from reading 0 index
+	muxHeapChangeMutex sync.Mutex
+}
+
+type muxableConnHeap []*connNode
+
+func (mh muxableConnHeap) Len() int { return len(mh) }
+
+func (mh muxableConnHeap) Less(i, j int) bool {
+	return mh[i].value < mh[j].value
+}
+
+func (mh muxableConnHeap) Swap(i, j int) {
+	mh[i], mh[j] = mh[j], mh[i]
+	mh[i].index = i
+	mh[j].index = j
+}
+
+func (mh *muxableConnHeap) Push(x interface{}) {
+	n := len(*mh)
+	cn := x.(*connNode)
+	cn.index = n
+	*mh = append(*mh, cn)
+}
+
+func (mh *muxableConnHeap) Pop() interface{} {
+	old := *mh
+	n := len(old)
+	cn := old[n-1]
+	cn.index = -1 // for safety
+	*mh = old[0 : n-1]
+	return cn
+}
+
+func (pool *connPool) getMuxableConn() <-chan ConnPoolContext {
+	cc := make(chan ConnPoolContext)
+
+	go func(cc chan<- ConnPoolContext) {
+		pool.muxHeapEmptyMutex.RLock()
+		cn := pool.muxHeap[0]
+		pool.muxHeapEmptyMutex.RUnlock()
+		cc <- cn
+	}(cc)
+
+	return cc
+}
+
+func (pool *connPool) deleteMuxableConn(cn *connNode) {
+	// lets us lazily get min in getMuxableConn
+	// without risking index out of bound
+	if atomic.AddInt64(&pool.numMuxableConns, -1) == 0 {
```

Comment:
this seems a bit brittle (e.g. what if we call `deleteMuxableConn` twice somewhere by accident?). Would it be better to use `CompareAndSwap` instead when you only flip a bit if you get empty (and the opposite in `addMuxableConn`)?

Reply:
Hmm. Well `deleteMuxableConn` is only called from https://github.com/doodles526/wormhole/blob/ad8a5792426a39367227fa87685f4dc91d1190de/net/conn_pool.go#L221 such that we check to ensure it isn't called twice. Only thing I suppose we could do is just collapse `deleteMuxableConn` into `delExistingConn` since deleteMuxableConn should only be called from one place

Reply:
👍